flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -5366,27 +5448,123 @@ class ArrayExprLowering {
+                      embox.slice().getDefiningOp())) {
+                // Perform a local rewrite of the embox to append the substr
+                // parameters.
+                auto trips = sliceOp.triples();
+                auto path = sliceOp.fields();
+                llvm::SmallVector<mlir::Value> typeparams = embox.typeparams();
+                auto newBounds = bounds;
+                if (newBounds.size() != 2) {
+                  if (charTy.hasConstantLen()) {
+                    auto idxTy = builder.getIndexType();
+                    newBounds.push_back(builder.createIntegerConstant(
+                        loc, idxTy, charTy.getLen()));
+                  } else {
+                    newBounds.push_back(typeparams.back());
+                  }
+                }
+                // Convert the lower bound to 0-based substring.
+                auto one = builder.createIntegerConstant(
+                    loc, newBounds[0].getType(), 1);
+                newBounds[0] =
+                    builder.create<mlir::SubIOp>(loc, newBounds[0], one);
+                // Convert the upper bound to a length.
+                newBounds[1] = builder.create<mlir::SubIOp>(
+                    loc, newBounds[1],
+                    builder.createConvert(loc, newBounds[1].getType(),
+                                          newBounds[0]));
+                auto newSlice =
+                    builder.create<fir::SliceOp>(loc, trips, path, newBounds);
+                auto memref = embox.memref();
+                auto shape = embox.shape();
+                if (auto off = mlir::dyn_cast_or_null<mlir::ConstantOp>(
+                        bounds[0].getDefiningOp()))
+                  if (auto len = mlir::dyn_cast_or_null<mlir::ConstantOp>(
+                          newBounds[1].getDefiningOp())) {
+                    // TODO: If the substring has constant arguments, move that
+                    // information into the type.
+                    (void)off;
+                    (void)len;
+                  }
+                auto charBoxTy = fir::BoxType::get(fir::SequenceType::get(
+                    fir::CharacterType::getUnknownLen(builder.getContext(),
+                                                      charTy.getFKind()),
+                    seqTy.getDimension()));
+                mlir::Value newEmbox = builder.create<fir::EmboxOp>(
+                    loc, charBoxTy, memref, shape, newSlice, typeparams);
+                embox.replaceAllUsesWith(newEmbox);
+                // FIXME: Should we be throwing away stuff in the old extended
+                // value?
+                return fir::BoxValue(newEmbox);
+              }
+              TODO(loc, "slice op expected");
+            }
+            TODO(loc, "unhandled box value");
+          },
+          [&](const auto &) -> fir::ExtendedValue {
+            TODO(loc, "unhandled substring base type");
```

Comment:
Given the constituent semantics is BoxValue for the base, shouldn't it be a fatal error to hit this ?